### ISSUE TITLE
Configurable spring precision

### ISSUE THREAD

**Issue by AndrewIngram** [chenglou/react-motion#100]
In a typical use case, we're using springs to position an element on the page. Depending on units, we don't need more than a few decimal places of precision to position elements accurately. At the moment, react-motion spits out far more frames than are necessary. I'm not sure exactly how we should do this, maybe something like this:

```javascript
let precisionFunc = (currentVal) => currentVal.toFixed(2);

<Spring endValue={{val, precision: precisionFunc, config: [70, 16]}}>
  {interpolated => {
    // stuff
  }
</Spring>
```

Then we could only update the spring's state if `precisionFunc(endValue)` has changed

**Comment by chenglou** [comment]
Does this have to be customizable? Seems like adjusting our error margin internally is good enough?

Also, we should render at least _once_ instead of just checking the error margin and decide that it's not worth moving half a pixel? You usually have the expectation that your render will be triggered (for e.g. side-effects and all) right?

**Comment by AndrewIngram** [comment]
Good point, I was thinking in terms of significant figures or decimal places, but error margin is more universally useful. Perhaps a error margin prop that defaults to 0.1%? I'd argue we don't want it bigger, since browsers are capable of non-integer positions, which make sense during animations.

**Comment by chenglou** [comment]
Percentage makes sense. But percentage relative to what? Can't be a value that might be 0.

**Comment by AndrewIngram** [comment]
Damned mathematics... This is why I was leaning towards a custom function, because the API consumer has a better notion of what's considered a big enough change to worry about.

**Comment by chenglou** [comment]
I think configurable error margin (even with a default) exposes too much API surface for the benefit it provides currently. Gotta find some other way of making this work.

**Comment by gre** [comment]
I don't know if this has been addressed yet but I'm quite in favor of the solution initially proposed.

IMO **user should have the control of precision grain** (e.g: integer, 2 digits, ...) because you don't know how an interpolated value is used at the end (it can be a style but it can also be displayed as a text right, or both. Or both, like in this pie chart I'm currently animating: <img height="20" src="https://cloud.githubusercontent.com/assets/211411/10564584/dc596e86-75b9-11e5-829f-6aa66182374f.png"> )

**Comment by nkbt** [comment]
I :+1: on this too. As you can see on http://nkbt.github.io/react-collapse/example/ (try to collapse) the element is not gone for a while, though visually it is definitely gone. I'd go for `0.0` or even `0.` precision in case of collapsing. It is not that easy to control this from outside. I mean it is fairly easy, but code gets some extra unnecessary baggage.

I'd love to have sort of `height: spring(100, {precision: 0})` API here.

**Comment by appsforartists** [comment]
I wrote a decorator that you can wrap around a `<Motion>`'s children to skip decimal values (presuming you are tweening from one pixel value to another:

```javascript
function memoizeRounded (callback) {
  var lastDict = {};
  var lastResult;

  return function (dict) {
    var changed = false;
    var roundedDict = {};

    for (var key in dict) {
      roundedDict[key] = Math.round(dict[key]) || dict[key];

      if (roundedDict[key] !== lastDict[key])
        changed = true;
    }

    if (!changed && lastResult) {
      return lastResult;

    } else {
      lastDict = roundedDict;
      return lastResult = callback(roundedDict);
    }
  }
}
```

I'll probably add it to [Gravel](https://github.com/appsforartists/gravel/tree/master/src) tomorrow.

**Comment by nkbt** [comment]
@appsforartists it should be really easy to add ability to specify precision if you replace `Math.round(dict[key])`. For many cases even for height, 0.1 precision is desirable. Sometimes 0.01. Depending on the kind of animation.

Wrapper solution is pretty straightforward indeed, but it would be much nicer to have it as part of spring() config.

**Comment by appsforartists** [comment]
Now published to npm:

 [`var memoizeRounded = require("gravel").helpers.memoizeRounded;`](https://github.com/appsforartists/gravel/blob/master/src/helpers/memoizeRounded.js)

**Comment by AndrewIngram** [comment]
So I was handling this in the child myself anyway, which for the most part was fine, but it didn't address the wasted renders of the Spring itself. There weren't any performance issues, but it did mean that when I was using React's Perf tools, the output showed those wasted renders, which somewhat obscures any other issues that might exist.

**Comment by chenglou** [comment]
This has now been fixed in v0.4.1 (newest release), under `spring(value, {stiffness: ..., damping: ..., precision: 0.01})`. Sorry for the long wait.

This is the first try at this and things can be changed. The current implementation is [really simple](https://github.com/chenglou/react-motion/blob/6a7a0f3f72610c6a7fdad2920b44e728ca6e172d/src/stepper.js#L33). It checks the newly passed-in `precision` to decide whether to stop the animation. So yes, you'll have to interpolate fewer frames. The default precision has been changed from `0.001` to `0.01`.

However, I didn't round the precision for the interpolating numbers themselves. The probability that none of your style's values changed at this frame because of rounding is really small. But if you have other arguments for why the numbers should be rounded please tell!

Last thing: as you can see from the code above, we stop when both the new velocity and the distance delta is smaller than `precision`. Of course there's no reason why they should share the same precision, but this is just to ease things up for now. Maybe this should be changed.

Closing this but glad to reopen/have a new issue to discuss about this!

Sorry again for the long wait.
